# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Buffer11.cpp
# Commit: 13282d7a47a5
# Full Hash: 13282d7a47a57a73be8d060efd9645b4d642f03e
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-09-08 09:46:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 3064469c073d -r 13282d7a47a5 gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Buffer11.cpp
--- a/gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Buffer11.cpp	Sat Sep 07 13:36:46 2019 +0000
+++ b/gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/Buffer11.cpp	Sat Sep 07 13:38:36 2019 +0000
@@ -391,7 +391,7 @@
             // TODO(jmadill): Use Context caps.
             if (offset == 0 && size >= mSize &&
                 size <= static_cast<UINT>(mRenderer->getNativeCaps().maxUniformBlockSize) &&
-                !mRenderer->getWorkarounds().useSystemMemoryForConstantBuffers)
+                !mRenderer->getFeatures().useSystemMemoryForConstantBuffers.enabled)
             {
                 ANGLE_TRY(getBufferStorage(context, BUFFER_USAGE_UNIFORM, &writeBuffer));
             }
@@ -557,16 +557,7 @@
 
 angle::Result Buffer11::markTransformFeedbackUsage(const gl::Context *context)
 {
-    BufferStorage *transformFeedbackStorage = nullptr;
-    ANGLE_TRY(getBufferStorage(context, BUFFER_USAGE_VERTEX_OR_TRANSFORM_FEEDBACK,
-                               &transformFeedbackStorage));
-
-    if (transformFeedbackStorage)
-    {
-        onStorageUpdate(transformFeedbackStorage);
-    }
-
-    invalidateStaticData(context);
+    ANGLE_TRY(markBufferUsage(context, BUFFER_USAGE_VERTEX_OR_TRANSFORM_FEEDBACK));
     return angle::Result::Continue;
 }
 
@@ -614,7 +605,7 @@
 bool Buffer11::canDeallocateSystemMemory() const
 {
     // Must keep system memory on Intel.
-    if (mRenderer->getWorkarounds().useSystemMemoryForConstantBuffers)
+    if (mRenderer->getFeatures().useSystemMemoryForConstantBuffers.enabled)
     {
         return false;
     }
@@ -628,6 +619,20 @@
     mIdleness[usage] = 0;
 }
 
+angle::Result Buffer11::markBufferUsage(const gl::Context *context, BufferUsage usage)
+{
+    BufferStorage *bufferStorage = nullptr;
+    ANGLE_TRY(getBufferStorage(context, usage, &bufferStorage));
+
+    if (bufferStorage)
+    {
+        onStorageUpdate(bufferStorage);
+    }
+
+    invalidateStaticData(context);
+    return angle::Result::Continue;
+}
+
 angle::Result Buffer11::garbageCollection(const gl::Context *context, BufferUsage currentUsage)
 {
     if (currentUsage != BUFFER_USAGE_SYSTEM_MEMORY && canDeallocateSystemMemory())
@@ -696,6 +701,12 @@
     return angle::Result::Continue;
 }
 
+angle::Result Buffer11::markRawBufferUsage(const gl::Context *context)
+{
+    ANGLE_TRY(markBufferUsage(context, BUFFER_USAGE_RAW_UAV));
+    return angle::Result::Continue;
+}
+
 angle::Result Buffer11::getRawUAVRange(const gl::Context *context,
                                        GLintptr offset,
                                        GLsizeiptr size,
@@ -704,14 +715,6 @@
     NativeStorage *nativeStorage = nullptr;
     ANGLE_TRY(getBufferStorage(context, BUFFER_USAGE_RAW_UAV, &nativeStorage));
 
-    BufferStorage *latestBuffer = nullptr;
-    ANGLE_TRY(getLatestBufferStorage(context, &latestBuffer));
-    // As UAVs could have been updated by the shader, they hold the latest version of the data.
-    if (latestBuffer != nativeStorage)
-    {
-        onStorageUpdate(nativeStorage);
-    }
-
     return nativeStorage->getRawUAV(context, static_cast<unsigned int>(offset),
                                     static_cast<unsigned int>(size), uavOut);
 }
@@ -949,13 +952,13 @@
 void Buffer11::initializeStaticData(const gl::Context *context)
 {
     BufferD3D::initializeStaticData(context);
-    onStateChange(context, angle::SubjectMessage::STORAGE_CHANGED);
+    onStateChange(angle::SubjectMessage::SubjectChanged);
 }
 
 void Buffer11::invalidateStaticData(const gl::Context *context)
 {
     BufferD3D::invalidateStaticData(context);
-    onStateChange(context, angle::SubjectMessage::STORAGE_CHANGED);
+    onStateChange(angle::SubjectMessage::SubjectChanged);
 }
 
 void Buffer11::onCopyStorage(BufferStorage *dest, BufferStorage *source)
@@ -1148,7 +1151,7 @@
     // Notify that the storage has changed.
     if (mOnStorageChanged)
     {
-        mOnStorageChanged->onStateChange(context, angle::SubjectMessage::STORAGE_CHANGED);
+        mOnStorageChanged->onStateChange(angle::SubjectMessage::SubjectChanged);
     }
 
     return angle::Result::Continue;